src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+	cbor_decode_trace();
+
+	if (result_len == 0 || *p_elem_count == 0) {
+		cbor_decode_trace();
+		(*pp_payload)--;
+		return false;
+	}
+
+	memset(p_result, 0, result_len);
+#ifdef CONFIG_BIG_ENDIAN
+	p_u8_result[result_len - 1] = additional;
+#else
+	p_u8_result[0] = additional;
+#endif /* CONFIG_BIG_ENDIAN */
+	if (additional > 23) {
+		uint32_t len = additional_len(additional);
+
+		if (len > result_len) {
+			cbor_decode_trace();
+			(*pp_payload)--;
+			return false;
+		}
+#ifdef CONFIG_BIG_ENDIAN
+		memcpy(&p_u8_result[result_len - len], *pp_payload, len);
+#else
+		for (uint32_t i = 0; i < len; i++) {
+			p_u8_result[i] = (*pp_payload)[len - i - 1];
+		}
+#endif /* CONFIG_BIG_ENDIAN */
+		(*pp_payload) += len;
+	}
+
+	if (*pp_payload > p_payload_end) {
+		/* Value extended beyond the payload. */
+		cbor_decode_trace();
+		(*pp_payload)--;
+		return false;
+	}
+	(*p_elem_count)--;
+	return true;
+}
+
+
+static bool int32_decode(uint8_t const **const pp_payload,
+			uint8_t const *const p_payload_end,
+			size_t *const p_elem_count, int32_t *p_result,
+			void *p_min_value, void *p_max_value)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, p_result,
+			4)) {
+		cbor_decode_trace();
+		return false;
+	}
+	if (*p_result < 0) {
+		/* Value is too large to fit in a signed integer. */
+		cbor_decode_trace();
+		return false;
+	}
+
+	if (major_type == CBOR_MAJOR_TYPE_NINT) {
+		*p_result *= -1;
+	}
+	if (!COMP_PTRS(int32_t, p_result, p_min_value, p_max_value)) {
+		cbor_decode_trace();
+		return false;
+	}
+	cbor_decode_print("val: %d\r\n", *p_result);
+	return true;
+}
+
+
+bool intx32_decode(uint8_t const **const pp_payload,
+			uint8_t const *const p_payload_end,
+			size_t *const p_elem_count, int32_t *p_result,
+			void *p_min_value, void *p_max_value)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (major_type != CBOR_MAJOR_TYPE_PINT
+		&& major_type != CBOR_MAJOR_TYPE_NINT) {
+		/* Value to be read doesn't have the right type. */
+		cbor_decode_trace();
+		return false;
+	}
+
+	if (!int32_decode(pp_payload, p_payload_end, p_elem_count,
+				p_result, p_min_value,
+				p_max_value)){
+		cbor_decode_trace();
+		return false;
+	}
+	return true;
+}
+
+
+static bool uint32_decode(uint8_t const **const pp_payload,
+			uint8_t const *const p_payload_end,
+			size_t *const p_elem_count,
+			void *p_result, void *p_min_value,
+			void *p_max_value)
+{
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, p_result,
+			4)) {
+		cbor_decode_trace();
+		return false;
+	}
+
+	if (!COMP_PTRS(uint32_t, p_result, p_min_value, p_max_value)) {
+		cbor_decode_trace();
+		return false;
+	}
+	cbor_decode_print("val: %u\r\n", *p_result);
+	return true;
+}
+
+
+bool uintx32_decode(uint8_t const **const pp_payload,
+		uint8_t const *const p_payload_end,
+		size_t *const p_elem_count,
+		uint32_t *p_result, void *p_min_value,
+		void *p_max_value)
+{
+	uint8_t major_type = **pp_payload >> 5;
```

Comment:
Comment the double-dereferenced bit switch

Reply:
Created macro for this. Macro is documented